packages/um/package.py

```diff
@@ -83,36 +166,108 @@ def setup_build_environment(self, env):
-
-        # Override some specific environment variables
-        env.set("optimisation_level", spec.variants["optim"].value)
-        env.set("platform_config_dir", spec.variants["platform"].value)
-        env.set("um_rev", f"vn{spec.version}")
-        components = ["casim", "jules", "shumlib", "socrates", "ukca"]
-        for comp in components:
-            env.set(f"{comp}_rev", f"um{spec.version}")
+                config_env[key] = config["env"][key].replace("\n=", "\n")
+
+        # Override the model UM revision based on the spec UM version.
+        key = "um_rev"
+        spec_um_rev = f"vn{spec.version}"
+        check_model_vs_spec(model, config_env, key, spec_um_rev)
+        config_env[key] = spec_um_rev
+
+        # Set DR_HOOK="false" until this package is available.
+        config_env["DR_HOOK"] = "false"
+
+        # Override those environment variables where a bool variant is specified.
+        bool_to_str = lambda b: "true" if b else "false"
+        for var in self._bool_variants:
+            spec_str_value = bool_to_str(spec.variants[var].value)
+            check_model_vs_spec(model, config_env, var, spec_str_value)
+            config_env[var] = spec_str_value
+
+        # Override those environment variables where an off/on variant is specified.
+        off_on_to_str = lambda off_on: "true" if off_on == "on" else "false"
+        for var in self._off_on_variants:
+            spec_value = spec.variants[var].value
+            if spec_value != "none":
+                spec_str_value = off_on_to_str(spec_value)
+                check_model_vs_spec(model, config_env, var, spec_str_value)
+                config_env[var] = spec_str_value
+
+        # Overide those environment variables where a revision variant is specified.
+        # If the variant is left unspecified, and the model does not specify a revision,
+        # then use a component revision based on the spec UM version.
+        for var in self._rev_variants:
+            spec_value = spec.variants[var].value
+            if spec_value != "none":
+                check_model_vs_spec(model, config_env, var, spec_value)
+                config_env[var] = spec_value
+            elif var not in config_env or config_env[var] == "":
+                config_env[var] = f"um{spec.version}"
+
+        # Override those environment variables where any other string variant is specified.
+        for var in self._other_variants:
+            spec_value = spec.variants[var].value
+            if spec_value != "none":
+                check_model_vs_spec(model, config_env, var, spec_value)
+                config_env[var] = spec_value
+
+        # Get the linker arguments for some dependencies.
         for fcm_libname in ["eccodes", "netcdf"]:
             linker_args = self._get_linker_args(spec, fcm_libname)
-            env.set(f"ldflags_{fcm_libname}_on", linker_args)
+            config_env[f"ldflags_{fcm_libname}_on"] = linker_args
```

Comment:
Could be done with one assignment.

Reply:
I left this as-is, as it is easy enough to understand what is happening here without combining the two lines into one assignment. Also, a single assignment would result in a 94 character line, which would be the longest line in the file, unless the line was somehow wrapped, continued or shortened.